Lib/gftools/builder/__init__.py

```diff
@@ -0,0 +1,312 @@
+            elif src.endswith("designspace"):
+                continue
+            else:
+                self.masters[src] = [Babelfont.open(src)]
+
+    def fill_config_defaults(self):
+        if "familyName" not in self.config:
+            self.logger.info("Deriving family name (this takes a while)")
+            self.load_masters()
+            familynames = set([x.info.familyName for x in self.masters.values()])
+            if len(familynames) != 1:
+                raise ValueError(
+                    "Inconsistent family names in sources (%s). Set familyName in config instead"
+                    % familynames
+                )
+            self.config["familyName"] = list(familynames)[0]
+        if "outputDir" not in self.config:
+            self.config["outputDir"] = "../fonts/"
+        if "vfDir" not in self.config:
+            self.config["vfDir"] = self.config["outputDir"] + "/variable/"
+        if "ttDir" not in self.config:
+            self.config["ttDir"] = self.config["outputDir"] + "/ttf/"
+        if "otDir" not in self.config:
+            self.config["otDir"] = self.config["outputDir"] + "/otf/"
+        if "woffDir" not in self.config:
+            self.config["woffDir"] = self.config["outputDir"] + "/webfonts/"
+
+        if "buildVariable" not in self.config:
+            self.config["buildVariable"] = True
+        if "buildStatic" not in self.config:
+            self.config["buildStatic"] = True
+        if "buildWebfont" not in self.config:
+            self.config["buildWebfont"] = self.config["buildStatic"]
+        if "autohintTTF" not in self.config:
+            self.config["autohintTTF"] = True
+        if "logLevel" not in self.config:
+            self.config["logLevel"] = "INFO"
+
+    def build_variable(self):
+        self.mkdir(self.config["vfDir"], clean=True)
+        args = {"output": ["variable"], "family_name": self.config["familyName"]}
+        all_variables = {}
+        for source in self.config["sources"]:
+            if not source.endswith(".designspace") and not source.endswith("glyphs"):
+                continue
+            self.logger.info("Creating variable fonts from %s" % source)
+            sourcebase = os.path.splitext(os.path.basename(source))[0]
+            args["output_path"] = os.path.join(
+                self.config["vfDir"], sourcebase + "-VF.ttf",
+            )
+            output_files = self.run_fontmake(source, args)
+            newname = self.rename_variable(output_files[0])
+            self.post_process(newname)
+            all_variables[source] = newname
+        if all_variables:
+            self.gen_stat(all_variables)
+
+    def run_fontmake(self, source, args):
+        if "output_dir" in args:
+            original_output_dir = args["output_dir"]
+            tmpdir = tempfile.TemporaryDirectory()
+            args["output_dir"] = tmpdir.name
+
+        if source.endswith(".glyphs"):
+            FontProject().run_from_glyphs(source, **args)
```

Comment:
I wouldn't recommend to use fontmake as a library. FontProject's constructor modifies logging global state (calls logging.basicConfig which will set a handler on the root logger the first time it gets called, then subsequently does nothing). I keep saying I want to change the API but never get to actually do it, I know..
Anyway, I think of fontmake as primarily a command line tool. It's "api" are its command line options, and the kind output it leaves behind on the filesystem. The "library" that does (or should do) the bulk of the business is ufo2ft (but I know there's still logic that's only available in fontmake itself, e.g. the fontmake.instantiator to interpolate static UFOs).
So I "reserve the right" to change the python interface of fontmake.font_project module without making any commitments its stability. In light of that, it would be advisable for your application as client of fontmake to treat it as a console tool and call it as a subprocess. But feel free to ignore me.

Reply:
FYI https://github.com/googlefonts/fontmake/issues/719

Reply:
OK, I hear what you're saying about the lack of API guarantees from `FontProject`. But I think it's worth the "risk". (Note also that Pilcrow takes the same approach of calling `FontProject` directly.)

There's a lot of logic in fontmake itself. Setting up the correct ufo2ft calls, connecting with glyphsLib, interpolating static UFOs... all of this is important stuff, and that's why it takes FontProject 1200 lines of code to do it. It's by no means a thin wrapper around ufo2ft.

I suppose we could retool to call `fontmake.__main__.main`. Perhaps we will do so when we need to. API guarantees are never *really* contracts - a new version of *anything* could change the API, and we would have to work around it. For now it works. If there are changes down the line, we deal with them.

Reply:
Yeah, doing that already.